daemon/policy.go

```diff
@@ -302,6 +304,79 @@ func (d *Daemon) PolicyAdd(rules policyAPI.Rules, opts *AddOptions) (newRev uint
 	return newRev, nil
 }
 
+type policyAddQueueOptions struct {
+	rules   policyAPI.Rules
+	opts    *AddOptions
+	context *policyQueueContext
+	reFetch bool
+}
+
+type policyQueueContext struct {
+	policyRevision uint64
+	err            error
+	done           chan bool
+	totalTime      spanstat.SpanStat
+	queueTime      spanstat.SpanStat
+}
+
+var policyAddQueueRequests = make(chan policyAddQueueOptions, 100)
+
+// policyAddToQueue adds the request to policyAdd in the queue and return a
+// context that will be updated when the done channel is closed.
+func (d *Daemon) policyAddToQueue(rules policyAPI.Rules, opts *AddOptions, reFetch bool) *policyQueueContext {
+	result := &policyQueueContext{
+		policyRevision: 0,
+		err:            nil,
+		done:           make(chan bool),
+	}
+	result.totalTime.Start()
+	result.queueTime.Start()
+
+	policyAddQueueRequests <- policyAddQueueOptions{
+		rules:   rules,
+		opts:    opts,
+		reFetch: reFetch,
+		context: result,
+	}
+
+	return result
+}
+
+// PolicyQueueWorker is a queue worker that taking care of PolicyAdd for
+// DNSProxy and Kubernetes watcher to avoid multiples races and lock all policy
+// repo.
+// In case that the channel event has a refetch the container, the rule will be
+// retrieved from policyRepo to make sure that the newest version is installed
+// correctly.
+func (d *Daemon) PolicyQueueWorker(ctx context.Context) {
+	log.Info("Initialize Policy Queue Worker")
+	for {
+		select {
+		case policyRequest, _ := <-policyAddQueueRequests:
+			policyRequest.context.queueTime.End(true)
+			var rules []*policyAPI.Rule
+			rules = policyRequest.rules
+			if policyRequest.reFetch {
+				newRules := policyAPI.Rules{}
+				d.policy.Mutex.Lock()
+				for _, sourceRule := range policyRequest.rules {
+					repoRules := d.policy.SearchRLocked(sourceRule.Labels)
+					newRules = append(newRules, repoRules...)
```

Comment:
Can you add a comment about why releasing the policy repository mutex is safe here if the rules that are added to the list are pointers?